src/mesh/mesh_function.C

```diff
@@ -512,7 +512,7 @@ void MeshFunction::gradient (const Point & p,
                   {
                     // local coordinates
                     for (std::size_t v=0; v<dim; v++)
-                      for (std::size_t xyz=0; xyz<dim; xyz++)
+                      for (std::size_t xyz=0; xyz<LIBMESH_DIM; xyz++)
```

Comment:
This looks like a good catch... we weren't getting MeshFunction gradients correct for lower-dimensional elements in higher-dimensional spaces?

Reply:
to be honest, I never tested. But I think, there would have been some error.
Be aware that this part is only reached for infinite elements. Since I didn't use it for their surfaces, no one did, I guess.